Add DEXseq wrapper

bgruening opened the pull request:
Inference of differential exon usage in RNA-Seq.

bgruening commented:
This PR should be ready for review. It will fix some outstanding cairo and R installation issues as described in this thread https://lists.galaxyproject.org/pipermail/galaxy-dev/2015-August/022659.html and related ones.

bgruening commented:
xref: https://github.com/galaxyproject/tools-devteam/pull/196 
ping @blankenberg

bgruening commented:
@yhoogstrate would you be willing to review this one? It's very similar to the DESeq2 one.

yhoogstrate commented:
Installation fails on multiple machines because package_cairo_1_12_14 misses glib:
``` fatal error: glib-object.h: No such file or directory #include <glib-object.h> ^ compilation terminated. make[4]```

I didn't find glib in the list of required packages (https://wiki.galaxyproject.org/Admin/Config/ToolDependenciesList), would it make sense to put it there?

yhoogstrate commented on tools/dexseq/dexseq.xml:
```diff
@@ -0,0 +1,181 @@
+        <requirement type="package" version="1.14">dexseq</requirement>
+    </requirements>
+    <code file="dexseq_helper.py" />
+    <stdio>
```
I got the following error ```Error: package ‘DESeq2’ required by ‘DEXSeq’ could not be found``` and it turned green. You could add this to solve it:
```
        <regex match="Error:"
            source="both"
            level="fatal"
            description="Error in the R script" />
```

yhoogstrate commented on tools/dexseq/dexseq.xml:
```diff
@@ -0,0 +1,181 @@
+
+**Inputs**
+
+DEXSeq_ takes count tables that generated from the dexseq_count as input. Count tables must be generated for each sample individually.
```
```
s/that//
```

or
```s/that/that are/```

yhoogstrate commented on tools/dexseq/dexseq_count.xml:
```diff
@@ -0,0 +1,110 @@
+    </inputs>
+
+    <outputs>
+        <data format="tabular" name="counts_file" label="DEXSeq count reads on ${on_string}">
```
Would it be possible to replace the ${on_string} to something that includes the original history name? History names usually include to which factor they belong. Now I've got a large list of files like "```24 DEXSeq count reads on data 19 and data 13```", which doesn't make it easier when I want to select them per factor in DEXSeq.

bgruening replied:
Not sure, {on_string} is the best what we have I think. This is a long-standing issue as far as I know.

bgruening commented:
Thanks @yhoogstrate I will work on this asap. The glib is a bummer, I need to think about this. I was not aware of this I thought glib is not installed inside of Docker.

yhoogstrate commented:
For DEXSeq-count I was using a GTF file that uses chromosomes with the names like *1, 2, 3* etc. while I was using a BAM file with chromosome names similar to: *chr1, chr2, chr3*. Then I was running these files in DEXSeq and at the end I got an error that it couldn't estimate dispersion. I know it's a dumb mistake, but maybe DEXSeq-count should throw an error if the sum of all counts is equal to zero if DEXSeq can't cope with them anyway?

I saw that DEXSeq was using 6 cores (this is hardcoded). I know DEXSeq is slower than e.g. DESeq or EdgeR, but I think 6 cores is more than most of the machines usually have. Should the variable not \${GALAXY_SLOTS:-6} be set in case a computer doesn't have 6 cores available?

In the tabular output Column 23 doesn't join lists. I get columns containing this: *c("ENST00000552564", "ENST00000293379")* but I think a comma delimited join would make more sense.

In the HTML output, once you click a gene, you can select "back	counts	expression	splicing	transcripts	results" of which only the links "back" and "results" are working. Also the colors from the legend are not visible in the table.

bgruening commented:
@yhoogstrate I have a problem with the glib dependency. I tried this in my Docker container and cairo is installing fine. I can see this in the log 'checking for glib ... no' and everything is still fine.

It is true that glib, without header files, is installed in the Docker container. I think it is a dependency of `pkg-config` which is a requirement of Galaxy.
So I don't know why you get this error. I have also no `glib-object.h` inside my container :(

yhoogstrate commented:
Hi @bgruening, 

I have tracked this issue down.
The ./configure.sh script searches for "gobject-2.0 glib-2.0 >= 2.14" (while they are not required to finish compilation successfully), using pkg-config indeed. When ./configure.sh finds glib-2-0.pc it gives "checking for glib ... yes" but compilation fails because this .pc file references to a missing lib.
After removing those *.pc files the cairo compilation succeeds again. Because I am not sure how often it happens that this package is giving problems (I had it on 2 machines and I can't remember I installed glibc from source deliberately), I tried to find an argument in ./configure.sh to disable glib permanently but I didn't succeed.

Excuse me for the unnecessairy work on your side...

bgruening commented:
> For DEXSeq-count I was using a GTF file that uses chromosomes with the names like *1, 2, 3* etc. while I was using a BAM file with chromosome names similar to: *chr1, chr2, chr3*. Then I was running these files in DEXSeq and at the end I got an error that it couldn't estimate dispersion because of the 0 values. I know it's a dumb mistake, but maybe DEXSeq-count should throw an error if the sum of all counts is equal to zero if DEXSeq can't cope with them anyway?

Good Idea, but I guess we should get this upstream?

> I saw that DEXSeq was using 6 cores (this is hardcoded). I know DEXSeq is slower than e.g. DESeq or EdgeR, but I think 6 cores is more than most of the machines usually have. Should the variable not \${GALAXY_SLOTS:-6} be set in case a computer doesn't have 6 cores available?

fixed

> In the tabular output Column 23 doesn't join lists. I get columns containing this: *c("ENST00000552564", "ENST00000293379")* but I think a comma delimited join would make more sense.

Upstream?

> In the HTML output, once you click a gene, you can select "back	counts	expression	splicing	transcripts results" of which only the links "back" and "results" are working. Also the colors from the legend are not visible in the table.

This should also be put upstream, isn't it?

yhoogstrate commented:
I will check for

> In the tabular output Column 23 doesn't join lists. I get columns containing this: c("ENST00000552564", "ENST00000293379") but I think a comma delimited join would make more sense.

because I think this one might be fixable within ```dexseq.R```. The others can be fixed upstream.

yhoogstrate commented:
Hi @bgruening ,

- The directory "dexseq/test_data" should be renamed to "dexseq/test-data" to get it working with ```planemo test```, otherwise it can't find the test data.

- The GTF format is set to ```ftype="bed"```, but this should be gtf. This also prevents ```planemo test``` from testing.

   * When those 2 items have been fixed, planemo test (0.14.1) crashes with the following error:
```
In DESeqDataSet(se, design, ignoreRank = TRUE) :
  some variables in design formula are characters, converting to factors
Error in unserialize(socklist[[n]]) : error reading from connection
Calls: estimateDispersions ... <Anonymous> -> recvOneData.SOCKcluster -> unserialize
```
(I got the same error with some of my own test data, I don't know what causes this.)
- ```<param name="countsFile" value="untreated1fb.txt,untreated2fb.txt,untreated2fb.txt,untreated3fb.txt"/>``` contains ```untreated2fb.txt``` twice, and ```<param name="countsFile" value="treated2fb.txt,treated3fb.txt,untreated3fb.txt,untreated4fb.txt"/>``` contains ```treated3fb.txt``` twice
- When I was running the test data from galaxy (web), it ends with the following error:
```
Fatal error: An undefined error occured, please check your intput carefully and contact your administrator.

... truncated ...

In DESeqDataSet(se, design, ignoreRank = TRUE) :
  some variables in design formula are characters, converting to factors
-- note: fitType='parametric', but the dispersion trend was not well captured by the
   function: y = a/x + b, and a local regression fit was automatically substituted.
   specify fitType='local' or 'mean' to avoid this message next time.
Error in DEXSeqHTML(res, path = opt$reportdir, FDR = opt$fdr, color = c("#B7FEA0",  : 
  There are no significant results in the test... nothing to report
```
It is possible this error is caused due to capturing the error I asked you to put in..

(I upgraded planemo to 0.15 and I don't get a HTML output anymore. When I have fixed this I will continue testing.)

yhoogstrate commented on tools/dexseq/dexseq.xml:
```diff
@@ -0,0 +1,185 @@
+                    <valid initial="string.letters,string.digits"><add value="_" /></valid>
+                </sanitizer>
+            </param>
+            <repeat name="rep_factorLevel" title="Factor level" min="2" max="2" default="2">
```
If this number is always two, it shouldn't be a repeat.

peterjc replied:
Certainly a fixed repeat of two is unusual, but it might still make sense compared to a literal repeat of a parameter (``factorLevel1`` and ``factorLevel2``) with repeats in the command tag too.

yhoogstrate replied:
The factor levels must be different to do a successful analysis, which is not the case because of the same default values of the repeat. It would be handier to have two separate inputs and set the values to some generic conditions e.g. one to "treated" and the other to "control".

yhoogstrate commented on tools/dexseq/dexseq.xml:
```diff
@@ -0,0 +1,185 @@
+                <param name="factorName" value="libtype"/>
+                <repeat name="rep_factorLevel">
+                    <param name="factorLevel" value="singleend"/>
+                    <param name="countsFile" ftype="tabular" value="treated1fb.txt,untreated1fb.txt,untreated2fb.txt"/>
```
```ftype="tabular"``` is only given here

yhoogstrate commented:
When I run planemo test, I get the following error:

```Error: nrow(sampleData) == ncol(countData) is not TRUE```

The error goes away when I disable 1 of the repeats of type ```<repeat name="rep_factorName">```

bgruening commented:
@yhoogstrate I removed the one repeat (we need to port this to DESeq2 as well) and fixed the tests. At least for me it's passing again.
Youri your last patch to join the R array does not work here.
I hope there are no other issues.

yhoogstrate commented:
@bgruening This should make the join work again. If you could test the latest version and it works, I think it can go in?

bgruening commented:
@yhoogstrate thanks! Awesome review! Now I need to update sailfish und we have a complete new RNA-seq stack.

yhoogstrate commented:
@bgruening It would be fun to have GOSeq too. It's a kind of GSEA but then for RNA-Seq.

bgruening commented:
@yhoogstrate if we can get this to work without requiring a internet connection I'm all in! :)

bgruening commented:
And uploaded to the TS/TTS!
